components/renku_data_services/connected_services/oauth_http.py

```diff
@@ -0,0 +1,611 @@
+    ) -> OAuthHttpError | models.AppInstallationList:
+        """Gets the users app installations if available."""
+        ...
+
+
+class OAuthHttpClientFactory(Protocol):
+    """Ways to create http-oauth clients."""
+
+    async def for_user_connection(self, user: APIUser, connection_id: ULID) -> OAuthHttpFactoryError | OAuthHttpClient:
+        """Create an oauth-http client for a given valid user and connection."""
+        ...
+
+    async def for_user_connection_raise(self, user: APIUser, connection_id: ULID) -> OAuthHttpClient:
+        """Same as `for_user_connection` but throws on error."""
+        ...
+
+    async def initiate_oauth_flow(
+        self, user: APIUser, provider_id: str, callback_url: str, next_url: str | None = None
+    ) -> OAuthHttpFactoryError | str:
+        """Create the authorization url to initiate the oauth flow.
+
+        Creates a connection in the database or resets its status to 'pending'.
+        """
+        ...
+
+    async def fetch_token(self, state: str, raw_url: str, callback_url: str) -> OAuthHttpFactoryError | OAuthHttpClient:
+        """Finishes the flow by trying to obtain a token given the response url from the authorization challenge."""
+        ...
+
+
+class _TokenCrypt(Protocol):
+    """Can encrypt/decrypt sensitive fields of the token set."""
+
+    def encrypt_token_set(self, token: dict[str, Any], user_id: str) -> models.OAuth2TokenSet:
+        """Encrypts sensitive fields of token set before persisting at rest."""
+        ...
+
+    def decrypt_token_set(self, token: dict[str, Any], user_id: str) -> models.OAuth2TokenSet:
+        """Decrypts sensitive fields of token set."""
+        ...
+
+
+class ConnectionLock:
+    """Lock updating a token."""
+
+    def __init__(self, session_maker: Callable[..., AsyncSession], id: ULID) -> None:
+        self._session_maker = session_maker
+        self._id = id
+
+    @asynccontextmanager
+    async def with_locked(self) -> AsyncGenerator[OAuth2ConnectionORM]:
+        """Lock the connection for update."""
+        async with self._session_maker() as session, session.begin():
+            result = await session.scalars(
+                sa.select(schemas.OAuth2ConnectionORM)
+                .where(schemas.OAuth2ConnectionORM.id == self._id)
+                .with_for_update(key_share=True)
+            )
+            conn = result.one()
+            yield conn
+            await session.flush()
+
+
+class _TokenCheck(_TokenCrypt, Protocol):
+    async def find_recently_updated_token(
+        self, connection_id: ULID, current_token: dict[str, Any], leeway: int
+    ) -> models.OAuth2TokenSet | None:
+        """Check the database for a recently updated token and return it."""
+        ...
+
+    def get_lock(self, id: ULID) -> ConnectionLock:
+        """Return a lock scoped by a connection id."""
+        ...
+
+
+class _SafeAsyncOAuthClient(AsyncOAuth2Client):  # type: ignore  # nosec: B107
```

Comment:
Nosec is not needed, and the type ignore can be more restrictive. Tested this with mypy locally. It works.

```suggestion
class _SafeAsyncOAuthClient(AsyncOAuth2Client):  # type: ignore[misc]
```